Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs

```diff
@@ -10,27 +10,64 @@ namespace UnityEngine.InputSystem.LowLevel
     /// A specialized event that contains the current IME Composition string, if IME is enabled and active.
     /// This event contains the entire current string to date, and once a new composition is submitted will send a blank string event.
     /// </summary>
-    [StructLayout(LayoutKind.Explicit, Size = InputEvent.kBaseEventSize + sizeof(int) + (sizeof(char) * kIMECharBufferSize))]
+    [StructLayout(LayoutKind.Explicit, Size = InputEvent.kBaseEventSize + sizeof(int) + (sizeof(char)))]
     public struct IMECompositionEvent : IInputEventTypeInfo
     {
-        // These needs to match the native ImeCompositionStringInputEventData settings
-        internal const int kIMECharBufferSize = 64;
-        public const int Type = 0x494D4553;
+        public const int Type = 0x494D4543;
 
         [FieldOffset(0)]
-        public InputEvent baseEvent;
+        internal InputEvent baseEvent;
 
         [FieldOffset(InputEvent.kBaseEventSize)]
-        public IMECompositionString compositionString;
+        internal int length;
+
+        [FieldOffset(InputEvent.kBaseEventSize + sizeof(int))]
+        internal char bufferStart;
 
         public FourCC typeStatic => Type;
 
-        public static IMECompositionEvent Create(int deviceId, string compositionString, double time)
+        internal IMECompositionString GetComposition()
         {
-            var inputEvent = new IMECompositionEvent();
-            inputEvent.baseEvent = new InputEvent(Type, InputEvent.kBaseEventSize + sizeof(int) + (sizeof(char) * kIMECharBufferSize), deviceId, time);
-            inputEvent.compositionString = new IMECompositionString(compositionString);
-            return inputEvent;
+            unsafe
+            {
+                fixed(char* buffer = &bufferStart)
+                {
+                    return new IMECompositionString(new IntPtr(buffer), length);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Queues up an IME Composition Event.!-- This event is for unit testing and debug purposes, and is slow, due to the variable way that IME Composition events work.
```

Comment:
Why is it slow?

Reply:
Removed, I didn't like the AllocHGlobal and copying 1 character at a time.  Had no good justification it was slow.